presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1876,22 +1965,56 @@ private Scope analyzeWith(Query node, Scope scope)
             return withScope;
         }
 
-        private void analyzeOrderBy(Query node, Scope scope)
+        private void analyzeOrderBy(Query node, Scope orderByScope)
+        {
+            checkState(node.getOrderBy().isPresent(), "orderBy is absent");
+
+            List<SortItem> sortItems = node.getOrderBy().get().getSortItems();
+            analyzeOrderBy(node, sortItems, orderByScope);
+        }
+
+        private List<Expression> analyzeOrderBy(QuerySpecification node, Scope orderByScope, List<Expression> outputExpressions)
         {
-            List<SortItem> items = node.getOrderBy()
-                    .map(OrderBy::getSortItems)
-                    .orElse(emptyList());
+            checkState(node.getOrderBy().isPresent(), "orderBy is absent");
+
+            List<SortItem> sortItems = node.getOrderBy().get().getSortItems();
+
+            if (node.getSelect().isDistinct()) {
+                verifySelectDistinct(node, outputExpressions);
+            }
+
+            return analyzeOrderBy(node, sortItems, orderByScope);
+        }
+
+        private void verifySelectDistinct(QuerySpecification node, List<Expression> outputExpressions)
+        {
+            for (SortItem item : node.getOrderBy().get().getSortItems()) {
+                Expression expression = item.getSortKey();
+
+                if (expression instanceof LongLiteral) {
+                    continue;
+                }
+
+                Expression rewrittenOrderByExpression = ExpressionTreeRewriter.rewriteWith(new OrderByExpressionRewriter(extractNamedOutputExpressions(node.getSelect())), expression);
+                int index = outputExpressions.indexOf(rewrittenOrderByExpression);
+                if (index == -1) {
```

Comment:
ORDER BY expressions and select expression are evaluated in different `ProjectNodes` as mentioned in: https://github.com/prestodb/presto/pull/7315/files#r101237523. Therefore
```
SELECT DISTINCT random() FROM (VALUES 1,2) t(a) ORDER BY random()
```
would return unordered (and possibly, but highly unlikely duplicated) rows. It is possible to address this issue, but some further work is needed. Also, we need to decide on semantics for non-deterministic functions in https://github.com/prestodb/presto/issues/7374. 
I would prefer to address the issue after this PR is merged and temporary disallow for distinct `QuerySpecifications` with nondeterministic ORDER BY. What do you think @martint ?